### ISSUE TITLE
Document public API surface before 1.0 (companion to #26 lint relaxation)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#33]
## Background

Issue #26 turned off the `missing_docs` CI gate (via `#![allow(missing_docs)]` at the crate root) because the crate is pre-1.0 and reshaping fast; forcing 144 doc comments onto a moving target would produce low-value filler that has to be rewritten when underlying APIs change. This issue is the deferred follow-up: document the public API surface once it has stabilized.

## Scope

At the time of #26's curation (commit `b39ba89`), `cargo +nightly doc --no-deps --all-features` produced 147 `missing documentation` warnings across 19 files. Top concentration:

- `src/policy/universal_inference.rs` — 40
- `src/env/games/snake/types.rs` — 16
- `src/buffer/rollout/storage.rs` — 11
- Top 9 files account for 88% of warnings.

Curator characterized the items as mostly real public API (serde wire formats, wasm-bindgen JS surface, training-script-tunable hyperparams) — i.e., users would benefit from docs. ~80% of fixes are mechanical (promoting existing `// [shape]` annotations to `///` doc comments).

## Three visibility tightenings (do these separately, not part of this issue)

Curator flagged 3 `pub` items in `src/optimize/{bayesian,pareto,scheduler}.rs` that appear to be empty placeholder structs. These should be `pub(crate)` or removed regardless of which path is chosen for #26. They have already been split off as a separate small follow-up.

## Acceptance

- Public API surface is documented enough that `#![allow(missing_docs)]` can be removed from `src/lib.rs` (or downgraded to `#![warn(missing_docs)]`).
- `cargo +nightly doc --no-deps --all-features` produces zero `missing documentation` warnings.
- CI gate restored.

## Trigger

Revisit when the API has been stable for a few weeks without major reshaping. Specific signals: no breaking trait widening, no major struct restructuring, no items moving between `pub` and `pub(crate)`.

## See also

- #26 — the lint-relaxation PR
- #19 / PR #24 — the `unresolved link` gate that #26 preserves

**Comment by rjwalters** [comment]
## Curator update (2026-05-14)

This issue remains **intentionally deferred**. Companion issue #26 was closed by commit `0a517e5`, which added `#![allow(missing_docs)]` to `src/lib.rs` rather than landing the 144 mechanical doc comments onto an actively-reshaping API. This is the deferred tracker — do NOT promote to `loom:issue` yet. A builder picking this up now would be doing throwaway work, because public-API surfaces are still moving (e.g., `MultiAgentEnvironment::step_multi` widened in #14, `JointPolicy::action_dims` added in #27, partial-rollout API added in #28).

The remainder of this comment refreshes the inventory and decomposes the work so a future builder (or human) does not need to re-explore.

---

### Refreshed warning inventory (post #14, #15, #27, #28)

`cargo +nightly doc --no-deps --no-default-features` with `#![allow(missing_docs)]` temporarily flipped to `#![warn(missing_docs)]`:

- **84 warnings** (no-default-features build)
- libtorch was not available in this environment, so the `--features training` count could not be re-measured directly. Based on the prior 147 count, training-only code paths contribute roughly **~60-65 additional warnings**.
- Current estimate of **total warnings: ~144-149**, essentially unchanged from the b39ba89 baseline of 147.
- Net effect of recent PRs:
  - #14 `step_multi` widening: docstring is already on the trait method (`src/multi_agent/environment.rs:65`) — no new warning.
  - #15 `JointMultiAgentTrainer`: code already documented during PR review.
  - #27 `MlpPolicy::action_dim`, `MultiDiscreteMlpPolicy::action_dims`: both have `///` doc comments — no new warnings.
  - #28 `RolloutBatch::from_buffer_partial`, `compute_advantages_partial`: both have `///` doc comments including `# Panics` sections — no new warnings. (Note: #28 chose `from_buffer_partial` over the alternative `get_filled_batch` API suggested in the issue.)

### Updated per-file concentration (no-default-features build)

| File | Warnings |
|---|---|
| `src/policy/universal_inference.rs` | 40 |
| `src/env/games/snake/types.rs` | 16 |
| `src/policy/inference.rs` | 14 |
| `src/inference/snake.rs` | 12 |
| `src/env/games/snake/snake.rs` | 1 |
| `src/env/games/snake.rs` | 1 |
| `src/lib.rs` | 1 (the lint-level note) |

These 4 files (`universal_inference.rs`, `snake/types.rs`, `inference.rs`, `inference/snake.rs`) account for **82 of 84** non-training warnings (98%). The training-feature top files from the b39ba89 inventory (`src/buffer/rollout/storage.rs:11`, plus 5 others) could not be re-counted here but are unlikely to have grown — #28's new API came with docs.

**Mechanical-promotion candidates** (`// [shape]` comments that can be `///`-ified verbatim or near-verbatim): ~32 across `policy/inference.rs` (14), `inference/snake.rs` (14), `universal_inference.rs` (4). The 18 shape/comment lines in `snake/types.rs` are mostly trivial enum variants and field names where the existing comment is just `// Snake ID` — those are sub-mechanical (a one-line `/// Up-direction` etc.).

### Visibility-tightening candidates (confirmed still placeholder stubs)

All three are still empty as of master HEAD `ea1622e`:

- `src/optimize/bayesian.rs`: `pub struct BayesianOptimizer;` + `// Placeholder for future Bayesian optimization implementation`
- `src/optimize/pareto.rs`: `pub struct ParetoFrontier;` + `// TODO: Implement Pareto frontier tracking`
- `src/optimize/scheduler.rs`: `pub struct TrialScheduler;` + `// TODO: Implement parallel trial execution with rayon`

**Recommendation**: split these into a tiny independent issue/PR. Two viable options:
1. Demote `pub` → `pub(crate)`. Cheapest; preserves the type if/when implementation lands.
2. Remove the structs entirely; reintroduce them when their impl modules are written.

Either is independent of #33's timing — no reason to keep visibility wrong even pre-1.0. This is a 5-minute change.

---

### Three-phase decomposition (for the future builder)

#### Phase 1 — Mechanical doc promotions (~1-2 hours, low risk)

Promote existing `// [shape]` annotations to `/// [shape]` doc comments. Affected files and counts:

- `src/policy/inference.rs` — 12 conv/fc weight+bias fields, 2 activation variants
- `src/inference/snake.rs` — 12 conv/fc weight+bias fields (sibling of the above)
- `src/policy/universal_inference.rs` — partial: ~4 struct-field shape comments

Pattern: each field already has an inline shape comment like `pub conv1_weight: Vec<Vec<Vec<Vec<f32>>>>, // [32, input_channels, 3, 3]`. Promote to:

```rust
/// Conv1 kernel weights, shape `[out_channels=32, in_channels=input_channels, kh=3, kw=3]`.
pub conv1_weight: Vec<Vec<Vec<Vec<f32>>>>,
```

Roughly 30-35 warnings cleared with no judgment calls.

#### Phase 2 — High-concentration files needing real prose (~4-8 hours)

These need actual writing, not just promotion:

1. **`src/policy/universal_inference.rs` (40 warnings)** — the largest concentration. Mix of:
   - `Activation` enum variants (`ReLU`, `Tanh`, `Sigmoid`, `Identity`, `Gelu`, `Swish`) — one-liners.
   - `Layer` enum variants (`Linear`, `Conv2d`, `Activation`, `Flatten`, `Reshape`, `Residual`) with struct-like fields — needs prose explaining wire format / serde semantics.
   - `InputSpec` / `OutputSpec` / `ModelMetadata` fields — these are the serde wire format for `.model.json` files; docs need to document the JSON schema, not just the Rust types.
   - These are real public API (consumed by wasm-bindgen JS callers and by training scripts that mutate metadata).
2. **`src/env/games/snake/types.rs` (16 warnings)** — `Direction` variants, `Position::{x, y}`, `Cell` variants (`Empty`, `Food`, `SnakeHead(usize)`, `SnakeBody(usize)`), `GameState` fields. Trivial prose but should call out the `usize` payload semantics (snake-id index into `scores` / `active_agents`).
3. **Training-feature top files** (re-measure with libtorch available): `src/buffer/rollout/storage.rs` (~11 from b39ba89), plus the other 5 from the prior top-9. The `RolloutBuffer` storage API is a real user-facing surface — docs should reference the `from_buffer_partial` semantics that #28 introduced.

#### Phase 3 — Cleanup (~30 min)

1. Delete the `#![allow(missing_docs)]` block at `src/lib.rs:16-20`.
2. Add `#![warn(missing_docs)]` (or `#![deny(missing_docs)]` if the project wants the CI gate).
3. Run `cargo +nightly doc --no-deps --features training` and confirm zero `missing documentation` warnings.
4. Confirm the Documentation CI job (`.github/workflows/ci.yml:115-120`, `RUSTDOCFLAGS="-D warnings"`) is green.
5. Update the comment in `src/lib.rs` (or remove the historical note) to reflect that the gate is now active.

---

### Triggers for revisiting (when is "stable enough"?)

Promote this issue to `loom:issue` **only when all of the following hold**:

1. **No breaking trait widening in the last 4 weeks.** Recent examples that would have failed this check: #14 (`step_multi` widened from `&[Self::Action]` to `&[Vec<i64>]`), #15 (`JointPolicy` trait introduced), #27 (`action_dims` added to `JointPolicy`).
2. **No `pub` items moved between `pub` and `pub(crate)` in the last 4 weeks** (a signal the API author is still deciding visibility).
3. **No new `pub` items added to the top-9 concentration files** without docs in the last 4 weeks (a signal that conventions for those modules are stable).
4. **The three `optimize/` placeholder structs** are either removed or have real implementations (so the Phase 1 mechanical sweep doesn't accidentally land docs on stubs that get deleted next week).
5. **A 1.0 release-prep milestone exists** (or the user explicitly opens the gate). Pre-1.0 churn during active port work (e.g., the slepian-wolf MARL port that just merged) is exactly the cost model that #26's `allow` is meant to absorb.

Light heuristic: when running `git log --oneline --since='4 weeks ago' src/` shows **no `pub` signature changes** to traits in `src/multi_agent/`, `src/policy/`, `src/buffer/`, the API has cooled enough.

---

### Acceptance criteria (final)

- [ ] All three optimize-module visibility tightenings landed (separate PR, independent of timing).
- [ ] Zero `missing documentation` warnings from `cargo +nightly doc --no-deps --features training` (and from `--no-default-features`).
- [ ] `#![allow(missing_docs)]` removed from `src/lib.rs`; `#![warn(missing_docs)]` (or deny) in place.
- [ ] Documentation CI job green on master.
- [ ] No regression in the `unresolved link` rustdoc gate from #19 / PR #24.

### Test plan

- `cargo +nightly doc --no-deps --no-default-features 2>&1 | grep -c 'missing documentation'` → must print `0`.
- `cargo +nightly doc --no-deps --features training 2>&1 | grep -c 'missing documentation'` → must print `0`.
- `cargo +nightly doc --no-deps --features training 2>&1 | grep -c 'unresolved link'` → must print `0` (regression guard for #19).
- CI Documentation job green on the resulting PR.

### See also

- #26 — the lint-relaxation PR (closed by `0a517e5`)
- #19 / PR #24 — the `unresolved link` gate that #26 preserves
- #14, #15, #27, #28 — recent reshaping evidence; these are the reason the deferral still makes sense

---

**Status**: `loom:curated`, NOT `loom:issue`. This is a future-execution tracker. Re-curate (or just re-run the inventory commands in this comment) when the "triggers for revisiting" above are satisfied.

**Comment by rjwalters** [comment]
Visibility cleanup completed in commit e51cea6: the three empty placeholder structs (`BayesianOptimizer`, `ParetoFrontier`, `TrialScheduler`) and their re-exports have been removed, along with the misleading module-level example calling `BayesianOptimizer::new(...)`. Each file keeps its module docstring + TODO note for when the real implementations land.

The remaining 84 `missing_docs` warnings are unchanged and still tracked here. Visibility-tightening checkbox done; no longer a prerequisite for the deferred doc-coverage work.
